File: core/src/main/java/com/xmartlabs/bigbang/core/helper/GeneralErrorHelper.java
package com.xmartlabs.bigbang.core.helper;

import android.support.annotation.NonNull;

import com.annimon.stream.Objects;
import com.annimon.stream.Stream;
import com.xmartlabs.bigbang.core.exception.EntityNotFoundException;
import com.xmartlabs.bigbang.core.helper.function.Consumer;

import java.net.ConnectException;
import java.net.UnknownHostException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CancellationException;

import javax.inject.Inject;

import io.reactivex.exceptions.CompositeException;
import lombok.Getter;
import timber.log.Timber;

/** Handles any {@link Throwable} thrown. */
@SuppressWarnings("unused")
public final class GeneralErrorHelper {
  private final List<Class<?>> UNTRACKED_CLASSES = Arrays.asList(
      CancellationException.class,
      ConnectException.class,
      EntityNotFoundException.class,
      UnknownHostException.class
  );
  private final StackTraceElement DUMMY_STACK_TRACE_ELEMENT =
      new StackTraceElement("General error logger", "", null, -1);
  private final Map<Class<? extends Throwable>, Consumer<? super Throwable>> throwableHandlers = new HashMap<>();

  @Inject
  public GeneralErrorHelper() {}

  @Getter
  private final Consumer<? super Throwable> generalErrorAction = t -> {
    if (t instanceof CompositeException) {
      CompositeException compositeException = (CompositeException) t;
      Stream.of(compositeException.getExceptions())
          .forEach(this::handleException);
    } else {
      handleException(t);
    }
  };

  private void handleException(@NonNull Throwable throwable) {
    if (!shouldHandleThrowable(throwable)
        || (throwable.getCause() != null && !shouldHandleThrowable(throwable.getCause()))) {
      if (!exceptionIsAlreadyBeingHandled(throwable)) {
        Timber.d(throwable, "Untracked exception");
      }
      markExceptionAsHandled(throwable);
      return;
    }

    if (throwableHandlers.containsKey(throwable.getClass())) {
      throwableHandlers.get(throwable.getClass()).accept(throwable);
    } else {
      Timber.e(throwable);
    }
    markExceptionAsHandled(throwable);
  }

  private boolean shouldHandleThrowable(@NonNull Throwable throwable) {
    return !exceptionIsAlreadyBeingHandled(throwable)
        && !UNTRACKED_CLASSES.contains(throwable.getClass());
  }

  private boolean exceptionIsAlreadyBeingHandled(@NonNull Throwable throwable) {
    return Stream.ofNullable(Arrays.asList(throwable.getStackTrace()))
        .anyMatch(value -> Objects.equals(value, DUMMY_STACK_TRACE_ELEMENT));
  }

  private void markExceptionAsHandled(Throwable throwable) {
    StackTraceElement[] stacktrace = Stream
        .concat(
            Stream.of(DUMMY_STACK_TRACE_ELEMENT),
            Stream.of(throwable.getStackTrace()))
        .toArray(StackTraceElement[]::new);
    throwable.setStackTrace(stacktrace);
  }

  /**
   * Add a new handler for a specific {@link Throwable} class.
   *
   * @param clazz   the {@link Throwable} class to be handled
   * @param handler the handler for the specified {@link Throwable}
   */
  public void setErrorHandlerForThrowable(Class<? extends Throwable> clazz, Consumer<? super Throwable> handler) {
    throwableHandlers.put(clazz, handler);
  }
}

File: core/src/main/java/com/xmartlabs/bigbang/core/helper/GeneralErrorHelper.kt
package com.xmartlabs.bigbang.core.helper

import com.xmartlabs.bigbang.core.exception.EntityNotFoundException
import com.xmartlabs.bigbang.core.extensions.andMatches
import com.xmartlabs.bigbang.core.extensions.toArray
import io.reactivex.exceptions.CompositeException
import timber.log.Timber
import java.net.ConnectException
import java.net.UnknownHostException
import java.util.concurrent.CancellationException
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.reflect.KClass

/** Handles any [Throwable] thrown.  */
@Singleton
class GeneralErrorHelper @Inject
constructor() {
  private val UNTRACKED_CLASSES: List<Class<out Exception>>
    get() = listOf(
        CancellationException::class.java,
        ConnectException::class.java,
        EntityNotFoundException::class.java,
        UnknownHostException::class.java
    )
  private val DUMMY_STACK_TRACE_ELEMENT = StackTraceElement("General error logger", "", null, -1)
  private val throwableHandlers = HashMap<KClass<out Throwable>, (Throwable) -> Unit>()

  val generalErrorAction: (Throwable) -> Unit = { t ->
    when(t) {
      is CompositeException -> t.exceptions.forEach(this::handleException)
      else -> handleException(t)
    }
  }

  private fun handleException(throwable: Throwable) {
    if (!shouldHandleThrowable(throwable) || throwable.cause.andMatches(this::shouldHandleThrowable)) {
      if (!exceptionIsAlreadyBeingHandled(throwable)) {
        Timber.d(throwable, "Untracked exception")
      }
    } else {
      if (throwableHandlers.containsKey(throwable::class)) {
        throwableHandlers[throwable::class]?.invoke(throwable)
      } else {
        Timber.e(throwable)
      }
    }
    markExceptionAsHandled(throwable)
  }

  private fun shouldHandleThrowable(throwable: Throwable) =
      !exceptionIsAlreadyBeingHandled(throwable) && !UNTRACKED_CLASSES.contains(throwable::class.java)

  private fun exceptionIsAlreadyBeingHandled(throwable: Throwable) =
      listOf(*throwable.stackTrace).any { it == DUMMY_STACK_TRACE_ELEMENT }

  private fun markExceptionAsHandled(throwable: Throwable) {
    throwable.stackTrace = mutableListOf(DUMMY_STACK_TRACE_ELEMENT)
        .apply { addAll(listOf(*throwable.stackTrace)) }
        .toArray()
  }

  /**
   * Add a new handler for a specific [Throwable] class.
   *
   * @param clazz   the [Throwable] class to be handled
   * *
   * @param handler the handler for the specified [Throwable]
   */
  fun setErrorHandlerForThrowable(clazz: KClass<out Throwable>, handler: (Throwable) -> Unit) =
      throwableHandlers.put(clazz, handler)
}
